### PR TITLE
Redirect inflation

### PR DESCRIPTION
**Pull request by tbfleming**
## Change Description
Create a contract to distribute inflation coming from the eosio contract action `setinflation`. The idea is to enable multiple accounts to receive inflation including `eosio.rex` via a new action `donatetorex`.

Backported from https://github.com/EOSIO/eosio.system/pull/53

## Deployment Changes
* [X] Deployment Changes

Deployment order:

1. Create `eosio.dist` (non-privileged) and install `eosio.distribute` on it.
1. Use the `setdistrib` action to set up a distribution.
1. Build the system contract with `USE_INFLATION_DISTRIBUTE`.
1. Update the system contract and abi.

Failure to complete steps 1 and 2 will cause `claimrewards` to abort if `USE_INFLATION_DISTRIBUTE` is enabled.

## API Changes
* [ ]  API Changes

## Documentation Additions
* [X] Documentation Additions

Add additional contract `eosio.distribute` to the docs.

### CODE REVIEW

**Comment by tbfleming** [comment]
What this does:
* The system contract, if built with `USE_INFLATION_DISTRIBUTE`, directs non-BP inflation to `eosio.dist` instead of `eosio.saving`.
* The system contract has a new `donatetorex` action.
* A new non-priviledged contract, `eosio.distribute`, redistributes incoming funds in a configurable way. The `setdistrib` action sets which accounts get which percentage.

What this allows:
* The rules for distributing non-BP inflation amoung multiple accounts are configurable instead of requiring future changes to the system contract.
* REX, which currently doesn't receive inflation, could receive inflation.

Additional notes:
* The system contract sends non-BP inflation to `eosio.dist` when BPs use the `claimrewards` action.
* If `eosio.distribute` is configured to send any inflation to REX, then this happens immediately.
* If `eosio.distribute` is configured to send funds to other accounts, then those funds are held. Accounts must use the `claimdistrib` action to receive them. This prevents a failure in a downstream contract from stopping `claimrewards`.
* `eosio.distribute` will build up unallocated funds over time from rounding dust.
* The distribution must be configured before updating the system contract. `claimrewards` will abort as long as there are no distributions set up.

File: contracts/eosio.distribute/src/eosio.distribute.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+namespace eosiodistribute {
+    using eosio::token;
+
+    distribute_contract::distribute_contract( name s, name code, datastream<const char*> ds )
+    :contract(s,code,ds),
+    _distrib_singleton(get_self(), get_self().value),
+    _claimers(get_self(), get_self().value)
+    {
+        _distrib_state = _distrib_singleton.exists() ? _distrib_singleton.get() : distribute_state{};
```
Comment (swatanabe) [inline contracts/eosio.distribute/src/eosio.distribute.cpp:12]: Isn't there a `get_or_default`?

File: contracts/eosio.distribute/src/eosio.distribute.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+   void distribute_contract::distribute(name from, name to, asset quantity, eosio::ignore<std::string> memo) {
+       if (to != get_self())
+           return;
+        check(quantity.symbol == eosiosystem::system_contract::get_core_symbol(), "Invalid symbol");
+        check(_distrib_state.accounts.size() > 0, "distribution accounts were not setup");
+
+        if(quantity.amount > 0) {
+            asset remaining = quantity;
```
Comment (swatanabe) [inline contracts/eosio.distribute/src/eosio.distribute.cpp:62]: Do we want to do anything with the remainder, like rolling it over to the next distribution?

File: contracts/eosio.distribute/src/eosio.distribute.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            for( const auto& acct : _distrib_state.accounts) {
+                if( remaining.amount == 0 ) {
+                    break;
+                }
+                asset dist_amount = quantity * acct.percent / max_distribute_pct;
+
+                if( remaining < dist_amount )
```
Comment (swatanabe) [inline contracts/eosio.distribute/src/eosio.distribute.cpp:69]: This shouldn't be possible unless the state table has been corrupted, right?

File: tests/eosio.distribute_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    // configure checks
+    BOOST_REQUIRE_EQUAL(t.error("missing authority of eosio.dist"), t.setdistrib(sys, {}));
+    BOOST_REQUIRE_EQUAL(t.wasm_assert_msg("Cannot set account to eosio.dist"), t.setdistrib(dist_acct, { {dist_acct, 10000} }));
+    BOOST_REQUIRE_EQUAL(t.wasm_assert_msg("Account does not exist: alice"), t.setdistrib(dist_acct, { {alice, 10000} }));
+    t.create_accounts_with_resources({alice, bob}, sys);
+    BOOST_REQUIRE_EQUAL(t.wasm_assert_msg("Total percentage does not equal 100%"), t.setdistrib(dist_acct, { {alice, 1000} }));
+    // force overflow
+    uint16_t pct = 65535;
+    // pct++;
+    BOOST_REQUIRE_EQUAL(t.wasm_assert_msg("Only positive percentages are allowed"), t.setdistrib(dist_acct, { {alice, ++pct} }));
```
Comment (swatanabe) [inline tests/eosio.distribute_tests.cpp:83]: The overflow has nothing to do with the contract's behavior.  The value that is serialized in the action data is 0.

**Comment by tbfleming** [comment]
Moved to https://github.com/eosnetworkfoundation/mandel-contracts/pull/3

**Pull request closed without merging**